build: cannot build with clang on windows

### Discussed in https://github.com/transmission/transmission/discussions/4893

<div type='discussions-op-text'>

<sup>Originally posted by **GaryElshaw** February 14, 2023</sup>
### What is the issue?

(Caveat: This is my second time building anything on Windows - ever)

So, the thing is, once you discover MSYS2, life starts looking up again. In case anyone else wants to set up an environment to build Transmission that excludes MS Visual Studio etc. https://www.freecodecamp.org/news/how-to-install-c-and-cpp-compiler-on-windows/. 

Still not there yet, but getting closer. Cmake gui can config and generate for MinGW Makefiles using Mingw64 gcc.exe and g++.exe compilers. But then a stumble.

In file included from C:/msys64/mingw64/include/winbase.h:19,
                 from C:/msys64/mingw64/include/windows.h:70,
                 from C:\Users\Gary\Documents\GitHub\transmission\libtransmission\error-types.h:10,
                 from C:\Users\Gary\Documents\GitHub\transmission\libtransmission\open-files.cc:16:
C:\Users\Gary\Documents\GitHub\transmission\libtransmission\open-files.cc: In function 'constexpr auto {anonymous}::isOpen(tr_sys_file_t)':
C:\Users\Gary\Documents\GitHub\transmission\libtransmission\file.h:40:25: error: 'reinterpret_cast' from integer to pointer
   40 | #define TR_BAD_SYS_FILE INVALID_HANDLE_VALUE
      |                         ^~~~~~~~~~~~~~~~~~~~
C:\Users\Gary\Documents\GitHub\transmission\libtransmission\open-files.cc:30:18: note: in expansion of macro 'TR_BAD_SYS_FILE'
   30 |     return fd != TR_BAD_SYS_FILE;
      |                  ^~~~~~~~~~~~~~~
make[2]: *** [libtransmission\CMakeFiles\transmission.dir\build.make:436: libtransmission/CMakeFiles/transmission.dir/open-files.cc.obj] Error 1
make[1]: *** [CMakeFiles\Makefile2:751: libtransmission/CMakeFiles/transmission.dir/all] Error 2
make: *** [Makefile:165: all] Error 2


--------------------------------------------------------------------------------------------------

Getting closer, but i really want to be building this without VS.

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	MSB3073	The command "setlocal
"C:\Program Files\CMake\bin\cmake.exe" -D TEST_TARGET=libtransmission-test -D TEST_EXECUTABLE=C:/Users/Gary/Documents/GitHub/transmission/build/tests/libtransmission/Debug/libtransmission-test.exe -D TEST_EXECUTOR= -D TEST_WORKING_DIR=C:/Users/Gary/Documents/GitHub/transmission/build/tests/libtransmission -D TEST_EXTRA_ARGS= -D TEST_PROPERTIES= -D TEST_PREFIX=LT. -D TEST_SUFFIX= -D TEST_FILTER= -D NO_PRETTY_TYPES=FALSE -D NO_PRETTY_VALUES=FALSE -D TEST_LIST=libtransmission-test_TESTS -D CTEST_FILE=C:/Users/Gary/Documents/GitHub/transmission/build/tests/libtransmission/libtransmission-test[1]_tests.cmake -D TEST_DISCOVERY_TIMEOUT=5 -D TEST_XML_OUTPUT_DIR= -P "C:/Program Files/CMake/share/cmake-3.25/Modules/GoogleTestAddTests.cmake"
if %errorlevel% neq 0 goto :cmEnd
:cmEnd
endlocal & call :cmErrorLevel %errorlevel% & goto :cmDone
:cmErrorLevel
exit /b %1
:cmDone
if %errorlevel% neq 0 goto :VCEnd
:VCEnd" exited with code 1.	libtransmission-test	C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets	159	
```
-----------------------------------------------------------------------------------------------

Using Cmake and Ninja, i got this error:

```
[91/294] Building CXX object libtransm...les/transmission.dir/open-files.cc.obj
FAILED: libtransmission/CMakeFiles/transmission.dir/open-files.cc.obj
C:\Strawberry\c\bin\c++.exe -DFMT_EXCEPTIONS=0 -DFMT_HEADER_ONLY=1 -DMINIUPNP_STATICLIB -DNATPMP_STATICLIB -DPACKAGE_DATA_DIR="\"C:/Program Files (x86)/transmission/share\"" -DSYSTEM_MINIUPNP -DWIDE_INTEGER_DISABLE_FLOAT_INTEROP -DWIDE_INTEGER_DISABLE_IOSTREAM -DWIN32 -DWITH_UTP -D_FILE_OFFSET_BITS=64 -D__TRANSMISSION__ -IC:/Users/Gary/Documents/GitHub/transmission/build/libtransmission -IC:/Users/Gary/Documents/GitHub/transmission/third-party/libutp/include -IC:/Users/Gary/Documents/GitHub/transmission/third-party/libb64/include -IC:/Users/Gary/Documents/GitHub/transmission/third-party/jsonsl -IC:/Users/Gary/Documents/GitHub/transmission/third-party/wildmat -isystem C:/Users/Gary/Documents/GitHub/transmission/build/third-party/deflate/include -isystem C:/dev/vcpkg/installed/x64-windows/include -isystem C:/Users/Gary/Documents/GitHub/transmission/cmake/../third-party/fast_float/include -isystem C:/Users/Gary/Documents/GitHub/transmission/build/third-party/psl/include -isystem C:/Users/Gary/Documents/GitHub/transmission/build/third-party/natpmp/include -isystem C:/Users/Gary/Documents/GitHub/transmission/build/third-party/miniupnpc/include -isystem C:/Users/Gary/Documents/GitHub/transmission/build/third-party/dht/include -isystem C:/Users/Gary/Documents/GitHub/transmission/cmake/../third-party/utfcpp/source -isystem C:/Users/Gary/Documents/GitHub/transmission/cmake/../third-party/wide-integer -isystem C:/Users/Gary/Documents/GitHub/transmission/cmake/../third-party/fmt/include -isystem C:/Users/Gary/Documents/GitHub/transmission/build/third-party/event/include -DWIN32 -DWINVER=0x0600 -D_WIN32_WINNT=0x0600 -DUNICODE -D_UNICODE -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_WINSOCK_DEPRECATED_NO_WARNINGS -D__USE_MINGW_ANSI_STDIO=1 -DWIN32_LEAN_AND_MEAN -DNOMINMAX -W -Wall -Wextra -Wcast-align -Wduplicated-cond -Wextra-semi -Wfloat-equal -Winit-self -Wint-in-bool-context -Wlogical-op -Wmissing-format-attribute -Wnull-dereference -Wpointer-arith -Wredundant-decls -Wrestrict -Wshadow -Wsign-compare -Wsuggest-override -Wuninitialized -Wunreachable-code -Wunused -Wunused-const-variable -Wunused-parameter -Wunused-result -Wwrite-strings -Wno-format -std=gnu++17 -MD -MT libtransmission/CMakeFiles/transmission.dir/open-files.cc.obj -MF libtransmission\CMakeFiles\transmission.dir\open-files.cc.obj.d -o libtransmission/CMakeFiles/transmission.dir/open-files.cc.obj -c C:/Users/Gary/Documents/GitHub/transmission/libtransmission/open-files.cc
In file included from C:/Strawberry/c/x86_64-w64-mingw32/include/winbase.h:19,
                 from C:/Strawberry/c/x86_64-w64-mingw32/include/windows.h:70,
                 from C:/Users/Gary/Documents/GitHub/transmission/libtransmission/error-types.h:10,
                 from C:/Users/Gary/Documents/GitHub/transmission/libtransmission/open-files.cc:16:
C:/Users/Gary/Documents/GitHub/transmission/libtransmission/open-files.cc: In function 'constexpr auto {anonymous}::isOpen(tr_sys_file_t)':
C:/Users/Gary/Documents/GitHub/transmission/libtransmission/file.h:40:25: error: reinterpret_cast from integer to pointer
 #define TR_BAD_SYS_FILE INVALID_HANDLE_VALUE
                         ^~~~~~~~~~~~~~~~~~~~
```

There are potentially a bunch of reasons for this, my inexperience developing on Windows is a good start, but i'm hoping someone can point me in the right direction.

### Which application of Transmission?

Qt app on Windows

### Which version of Transmission?

4.0 git</div>

--- a/libtransmission/open-files.cc
+++ b/libtransmission/open-files.cc
@@ -25,7 +25,7 @@
 namespace
 {
 
-[[nodiscard]] auto constexpr isOpen(tr_sys_file_t fd) noexcept
+[[nodiscard]] auto isOpen(tr_sys_file_t fd) noexcept
 {
     return fd != TR_BAD_SYS_FILE;
 }

